clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/view/EcsServerClusterProvider.java

```diff
@@ -0,0 +1,269 @@
+
+  @Autowired
+  public EcsServerClusterProvider(AccountCredentialsProvider accountCredentialsProvider, AmazonClientProvider amazonClientProvider, ContainerInformationService containerInformationService) {
+    this.accountCredentialsProvider = accountCredentialsProvider;
+    this.amazonClientProvider = amazonClientProvider;
+    this.containerInformationService = containerInformationService;
+    getClusters();
+  }
+
+  @Override
+  public Map<String, Set<EcsServerCluster>> getClusters() {
+    Map<String, Set<EcsServerCluster>> clusterMap = new HashMap<>();
+
+    for (AccountCredentials credentials: accountCredentialsProvider.getAll()) {
+      if (credentials instanceof AmazonCredentials) {
+        clusterMap = findClusters(clusterMap, (AmazonCredentials) credentials);
+      }
+    }
+    return clusterMap;
+  }
+
+  private Map<String, Set<EcsServerCluster>> findClusters(Map<String, Set<EcsServerCluster>> clusterMap,
+                                                          AmazonCredentials credentials) {
+    for (AmazonCredentials.AWSRegion awsRegion: credentials.getRegions()) {
+      clusterMap = findClustersForRegion(clusterMap, credentials, awsRegion);
+    }
+
+    return clusterMap;
+  }
+
+  private Map<String, Set<EcsServerCluster>> findClustersForRegion(Map<String, Set<EcsServerCluster>> clusterMap,
+                                                                   AmazonCredentials credentials,
+                                                                   AmazonCredentials.AWSRegion awsRegion) {
+
+    AmazonECS amazonECS = amazonClientProvider.getAmazonEcs(credentials.getName(), credentials.getCredentialsProvider(), awsRegion.getName());
+    AmazonEC2 amazonEC2 = amazonClientProvider.getAmazonEC2(credentials.getName(), credentials.getCredentialsProvider(), awsRegion.getName());
+    AmazonElasticLoadBalancing amazonELB = amazonClientProvider.getAmazonElasticLoadBalancingV2(credentials.getName(), credentials.getCredentialsProvider(), awsRegion.getName());
+
+    for (String clusterArn: amazonECS.listClusters().getClusterArns()) {
+      String ecsClusterName = inferClusterNameFromClusterArn(clusterArn);
+
+      ListServicesResult result = amazonECS.listServices(new ListServicesRequest().withCluster(ecsClusterName));
+      for (String serviceArn: result.getServiceArns()) {
+
+        ServiceMetadata metadata = extractMetadataFromServiceArn(serviceArn);
+        Set<Instance> instances = new HashSet<>();
+
+        DescribeLoadBalancersResult loadBalancersResult = amazonELB.describeLoadBalancers(new DescribeLoadBalancersRequest());
+        Set<com.netflix.spinnaker.clouddriver.model.LoadBalancer> loadBalancers = extractLoadBalancersData(loadBalancersResult);
+
+        ListTasksResult listTasksResult = amazonECS.listTasks(new ListTasksRequest().withServiceName(serviceArn).withCluster(ecsClusterName));
+        if (listTasksResult.getTaskArns() != null && listTasksResult.getTaskArns().size() > 0) {
```

Comment:
I do not think that getTaskArns() can return null

Nitpicking: `!listTaskResult.getTaskArns().isEmpty`

also, I do noth think that you need to store the listTasksResult, you only use the getTaskArns. Why don't you just store the result of that call?

Same might apply to the other *Result variables on this method.

Reply:
A service can not yet have tasks, or be scaled down to 0 tasks.  As for the list being null, it depends on the AWS implementation which can change over time.  I'd rather be cautious here

Reply:
As for storing the result of the call - it's a style preference whether you do it the way you suggested, or the way I did it.  It's easier to understand code through debugging it step by step when you store the result like I did